src/main/scala/ignition/core/utils/S3Utils.scala

```diff
@@ -0,0 +1,91 @@
+package ignition.core.utils
+
+import com.amazonaws.auth.EnvironmentVariableCredentialsProvider
+import com.amazonaws.services.s3.AmazonS3Client
+import com.amazonaws.services.s3.model.{S3ObjectSummary, ObjectListing}
+import ignition.core.jobs.utils.PathDateExtractor
+import ignition.core.utils.DateUtils._
+import org.joda.time.DateTime
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+import scala.util.Try
+
+object S3Utils {
+
+  implicit lazy val amazonS3ClientFromEnvironmentVariables = new AmazonS3Client(new EnvironmentVariableCredentialsProvider())
+
+  def s3List(bucket: String, prefix: String, predicate: S3ObjectSummary => Boolean = _ => true)
+            (implicit s3: AmazonS3Client): List[S3ObjectSummary] = {
+    def inner(acc: mutable.ArrayBuffer[S3ObjectSummary], listing: ObjectListing): List[S3ObjectSummary] = {
+      acc ++= listing.getObjectSummaries.toList.filter(predicate)
+      if (listing.isTruncated)
+        inner(acc, s3.listNextBatchOfObjects(listing))
+      else
+        acc.toList
+    }
+
+    inner(new mutable.ArrayBuffer[S3ObjectSummary], s3.listObjects(bucket, prefix))
+  }
+
+  def s3ListAndFilterFiles(bucket: String,
+                           prefix: String,
+                           start: Option[DateTime] = None,
+                           end: Option[DateTime] = None,
+                           endsWith: Option[String] = None,
+                           exclusionPattern: Option[String] = Option("_$folder$"),
+                           predicate: S3ObjectSummary => Boolean = _ => true)
+                          (implicit s3: AmazonS3Client, pathDateExtractor: PathDateExtractor): List[S3ObjectSummary] = {
+
+    def excludePatternValidation(s3Object: S3ObjectSummary, exclusionPatternOption: Option[String]): Option[S3ObjectSummary] =
```

Comment:
I don't know if it helps. But I think that most of this code could be simplified by filters (S3ObjectSummary) => Boolean